controller/src/controller.rs

```diff
@@ -0,0 +1,149 @@
+    async fn search_for_node_to_update(&self) -> Option<BottlerocketNode> {
+        for brn in self.all_nodes() {
+            // If we determine that the spec should change, this node is a candidate to begin updating.
+            let next_spec = determine_next_node_spec(&brn);
+            if next_spec != brn.spec {
+                match self.progress_node(brn.clone()).await {
+                    Ok(_) => return Some(brn),
+                    Err(_) => {
+                        // Errors connecting to the k8s API are ignored (and also logged by `progress_node()`).
+                        // We'll just move on and try a different node.
+                        continue;
+                    }
+                }
+            }
+        }
+        None
+    }
+
+    /// Runs the event loop for the Brupop controller.
+    ///
+    /// Because the controller wants to gate the number of simultaneously updating nodes, we can't allow the update state machine
+    /// of each individual bottlerocket node to run concurrently and in an event-driven fashion, as is typically done with controllers.
+    /// Instead, we will keep an updated store of `BottlerocketNode` objects based on cluster events, and then periodically make
+    /// scheduling decisions based on that store.
+    ///
+    /// The controller is designed to run on a single node in the cluster and rely on the scheduler to ensure there is always one
+    /// running; however, it could be expanded using leader-election and multiple nodes if the scheduler proves to be problematic.
+    pub async fn run(&self) -> Result<()> {
+        // On every iteration of the event loop, we reconstruct the state of the controller and determine its
+        // next actions. This is to ensure that the operator would behave consistently even if suddenly restarted.
+        loop {
+            // Brupop typically only operates on a single node at a time. Here we find the set of nodes which is currently undergoing
+            // change, to ensure that errors resulting in multiple nodes changing state simultaneously is not unrecoverable.
+            let active_set = self.active_node_set();
+            event!(Level::TRACE, ?active_set, "Found active set of nodes.");
+
+            if !active_set.is_empty() {
+                // Try to push forward all active nodes, gathering results along the way.
+                let mut nodes: Vec<BottlerocketNode> = active_set.into_values().collect();
+
+                for brn in nodes.drain(..) {
+                    // Timeouts and errors are logged by instrumentation in `progress_node()`.
+                    #[allow(unused_must_use)]
+                    {
+                        self.progress_node(brn).await;
+                    }
+                }
+            } else {
+                // If there's nothing to operate on, check to see if any other nodes are ready for action.
+                let new_active_node = self.search_for_node_to_update().await;
```

Comment:
Shouldn't this be executed always? Regardless of the active nodes being empty?

Reply:
We want to restrict brupop to only operate on a single node at a time. Even if other nodes are ready to become active, we want to wait until no other nodes are active.